app-ng/src/job.lisp

```diff
@@ -0,0 +1,145 @@
+(in-package #:qvm-app-ng)
+
+(adt:defdata job-status
+  job-status-fresh
+  job-status-running
+  job-status-killed
+  (job-status-finished t)
+  (job-status-error condition))
+
+(defun job-status-name (job)
```

Comment:
Is this unsafe? In what way could this result in a bad result or bad access?

Reply:
It's unsafe if some other thread writes the status concurrently, no? In general I'd think that anything that reads/writes job slots is unsafe unless it's holding the lock. By "unsafe" I really mean "racy", but maybe also "unsafe" in the sense you might get garbage data back if writing to the slot value isn't atomic.

`job-peek` is one where it probably doesn't make sense unless you're already holding the lock, since the whole reason you would "peek" is most likely because you're about to do something with the job conditional on the result.

Reply:
I _think_ this is OK. I don't think it's possible to get a bad answer, just an invalid answer by the time the call ends, but that'll be true even with a lock

Reply:
Agreed, the caller can always acquire the lock themselves first when they need consistent results.

I still suspect that it's at least theoretically possible to get a torn read here, but agree that it should be unlikely if not impossible on many (most?) implementations. On SBCL at least, a slot write always compiled down to a single `MOV` for the handful of cases I tried, which mostly puts my mind at ease.